File: backend/app/documents/csa.py
"""Chat-driven field extraction for the Cloud Service Agreement.

Scope note: the real template also defines advanced/rarely-used terms
(Increased Cap Amount, Increased Claims, Unlimited Claims, Additional
Warranties, an optional DPA reference) that this chat does not ask about.
Those substitute to fixed defaults in the frontend's substitution map rather
than becoming extra chat questions, matching the depth already used for the
Mutual NDA.
"""

import datetime

from pydantic import BaseModel

from app.chat import ChatMessage
from app.chat import run_chat_turn as _run_chat_turn
from app.documents.common import CHAT_RULES, PartyFields

SLUG = "csa"
CATALOG_NAME = "Cloud Service Agreement"
TEMPLATE_FILE = "CSA.md"
DOCUMENT_TITLE = "Cloud Service Agreement"


class CsaChatFields(BaseModel):
    effectiveDate: str | None = None
    orderDate: str | None = None
    subscriptionPeriod: str | None = None
    nonRenewalNoticeDate: str | None = None
    technicalSupport: str | None = None
    paymentProcess: str | None = None
    useLimitations: str | None = None
    governingLaw: str | None = None
    chosenCourts: str | None = None
    generalCapAmount: str | None = None
    customer: PartyFields | None = None
    provider: PartyFields | None = None


class CsaChatTurn(BaseModel):
    reply: str
    fields: CsaChatFields
    done: bool


SYSTEM_PROMPT = f"""You are an assistant helping a user fill in a Cloud Service Agreement (CSA),
for a provider to license a cloud-hosted product to a customer.

Today's date is {datetime.date.today().isoformat()}.

Ask about the following fields, ONE QUESTION AT A TIME, in this fixed order:
1. effectiveDate - the date the Framework Terms start (always output as YYYY-MM-DD)
2. orderDate - the date this specific Order Form starts (YYYY-MM-DD)
3. subscriptionPeriod - the length of the subscription (e.g. "1 year")
4. nonRenewalNoticeDate - how far before renewal a party must give notice not to renew
   (e.g. "30 days before the end of the Subscription Period")
5. technicalSupport - a short description of the support provided
6. paymentProcess - how invoicing/payment works (e.g. "invoiced monthly in advance")
7. useLimitations - any limits on how the customer may use the product, or "None"
8. governingLaw and chosenCourts
9. generalCapAmount - the total liability cap for either party (e.g. a dollar amount)
10. customer: name, title, company, and noticeAddress - ask for these together as one question
11. provider: name, title, company, and noticeAddress - ask for these together as one question
{CHAT_RULES}"""


def run_chat_turn(messages: list[ChatMessage]) -> CsaChatTurn:
    return _run_chat_turn(SYSTEM_PROMPT, messages, CsaChatTurn)

File: frontend/lib/documents/csa.ts
import type { DocumentModule } from "./types";
import {
  blankParty,
  formatLongDate,
  getParty,
  mergeSubstitutions,
  roleAliases,
  standardLiabilityDefaults,
  textValue,
} from "./common";

export const csa: DocumentModule = {
  slug: "csa",
  catalogName: "Cloud Service Agreement",
  templateFile: "CSA.md",
  documentTitle: "Cloud Service Agreement",
  subtitle: "Common Paper Cloud Service Agreement Standard Terms Version 2.1",
  defaultValues: {
    effectiveDate: "",
    orderDate: "",
    subscriptionPeriod: "",
    nonRenewalNoticeDate: "",
    technicalSupport: "",
    paymentProcess: "",
    useLimitations: "None.",
    governingLaw: "",
    chosenCourts: "",
    generalCapAmount: "",
    customer: { ...blankParty },
    provider: { ...blankParty },
  },
  groups: [
    {
      title: "Key Terms",
      fields: [
        { label: "Effective Date", render: (v) => formatLongDate(v.effectiveDate) },
        { label: "Governing Law", key: "governingLaw" },
        { label: "Chosen Courts", key: "chosenCourts" },
        { label: "General Cap Amount", key: "generalCapAmount" },
      ],
    },
    {
      title: "Order Form",
      fields: [
        { label: "Order Date", render: (v) => formatLongDate(v.orderDate) },
        { label: "Subscription Period", key: "subscriptionPeriod" },
        { label: "Non-Renewal Notice Date", key: "nonRenewalNoticeDate" },
        { label: "Technical Support", key: "technicalSupport" },
        { label: "Payment Process", key: "paymentProcess" },
        { label: "Use Limitations", key: "useLimitations" },
        { label: "Customer", key: "customer", party: true },
        { label: "Provider", key: "provider", party: true },
      ],
    },
  ],
  buildSubstitutions: (values) => {
    const customer = getParty(values, "customer");
    const provider = getParty(values, "provider");
    return mergeSubstitutions(
      {
        keyterms_link: {
          "Effective Date": formatLongDate(values.effectiveDate),
          "Governing Law": textValue(values, "governingLaw", "[Governing Law]"),
          "Chosen Courts": textValue(values, "chosenCourts", "[Chosen Courts]"),
          "General Cap Amount": textValue(values, "generalCapAmount", "[General Cap Amount]"),
        },
        orderform_link: {
          "Order Date": formatLongDate(values.orderDate),
          "Subscription Period": textValue(values, "subscriptionPeriod", "[Subscription Period]"),
          "Subscription Periods": textValue(values, "subscriptionPeriod", "[Subscription Period]"),
          "Non-Renewal Notice Date": textValue(
            values,
            "nonRenewalNoticeDate",
            "[Non-Renewal Notice Date]",
          ),
          "Technical Support": textValue(values, "technicalSupport", "[Technical Support]"),
          "Payment Process": textValue(values, "paymentProcess", "[Payment Process]"),
          "Use Limitations": textValue(values, "useLimitations", "None."),
        },
      },
      { keyterms_link: standardLiabilityDefaults("Provider", "Customer") },
      { coverpage_link: roleAliases("Customer", customer) },
      { coverpage_link: roleAliases("Provider", provider) },
    );
  },
};
